Add parameter osc_freq in tests and in tutorial

merav-aharoni opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
In the previous version, all tests ignored osc_freq.


### Details and comments

Fixes #176.

merav-aharoni commented:
Ready for review.

yaelbh commented on test/test_t2ramsey.py:
```diff
@@ -227,10 +225,11 @@ def test_t2ramsey_parallel(self):
         estimated_freq = [0.1, 0.12]
         delays = [list(range(1, 60)), list(range(1, 50))]
-        dt_factor = 1
+        dt_factor = 1e-6
```
What's the origin of this change?

merav-aharoni replied:
I added this just to make the numbers more realistic (microsec rather than sec)

nkanazawa1989 commented:
Can we call this just `freq`?

For example, in resonance analysis
https://github.com/Qiskit/qiskit-experiments/blob/70d9b487dafe78dd9192647af623994fd2f36698/qiskit_experiments/characterization/resonance_analysis.py#L85

yaelbh commented on qiskit_experiments/characterization/t2ramsey.py:
```diff
@@ -80,7 +80,9 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
             circ.h(0)
             circ.delay(delay, 0, self._unit)
-            circ.p(2 * np.pi * self._osc_freq, 0)
+            nosc = delay / self._delays[0]
```
Will appreciate if someone with more experience with Ramsey experiments checks this line

merav-aharoni replied:
@nkanazawa1989 , can you have a look at this line? the origin is from Ignis. The idea is that the rotation in the phase gate increase in the same proportion as the delay.

yaelbh reviewed (commented):
1. The fitted frequency is the sum of the oscillation frequency and the detuning frequency. Is it what experimentalists expect to see?
2. Did you verify that this gives reasonable results on a real device? 
3. The tutorial says that the fitted frequency is the detuning frequency, I guess that this sentence has to be modifed.
4. In the fake backend there is some confusion, the parameters (except for t2ramsey) are called "guesses", although there are no guesses involved, only parameters to simulate.
5. In the tutorial, the frequency which is the sum of the oscillation frequency and the detuning frequency is 0.2, whereas the oscillation frequency is larger (0.1111). I guess that it's possible if the detuning frequency is negative, but I also guess that you'll want to revisit this example.
6. The tutorial doesn't explain what's osc_freq.
7. In the tutorial, Cell 6: there's a line in comment, and p0 is eventually not used (which is different from the text).
8. Many cells in the tutorial don't set osc_freq, you may want to revisit it.

merav-aharoni commented:
Regarding your comment, @nkanazawa1989 , I would rather leave it as `osc_freq` because `f` or `freq` is used for the the fitted frequency. I improved the explanation in the tutorial. It would be great if you could look at the explanation and see if it is correct and clear enough.

merav-aharoni commented:
@yael, here are responses to your long comment - for some reason I could not answer directly below yours.
1. I didn't get an answer on what the users expect. For now, I wrote a better explanation at the beginning of the tutorial, so I hope that will suffice. We can change later, if we get such feedback.
2. Results on the devices are not as nice. It could be that my parameters need more tuning.
3. See my answer on 1.
4. Removed the word `guess` from all parameter names.
5. I fixed the values of 'osc_freq` and 'f` as you pointed out.
6. See 1.
7. I couldn't find this - I probably corrected it previously.
8. Added osc_freq to all examples.

chriseclectic commented on qiskit_experiments/library/characterization/t2ramsey.py:
```diff
@@ -80,7 +80,9 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
             circ.h(0)
             circ.delay(delay, 0, self._unit)
-            circ.p(2 * np.pi * self._osc_freq, 0)
+            nosc = delay / self._delays[0]
```
Could `_delays[0]` be 0? if so this could raise an error

merav-aharoni replied:
1. Added a check that `delays` are strictly increasing.
2. If delay[0] is 0 then take delay[1] as min value for division to avoid the error you pointed out.

merav-aharoni commented:
@chriseclectic , I believe this PR is in the category of bug fixes, so should be included in the release. Please review and merge.

chriseclectic commented on qiskit_experiments/library/characterization/t2ramsey.py:
```diff
@@ -103,20 +103,26 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
                 dt_factor = getattr(backend._configuration, "dt")
             except AttributeError as no_dt:
                 raise AttributeError("Dt parameter is missing in backend configuration") from no_dt
+        if not all(d1 < d2 for d1, d2 in zip(self._delays, self._delays[1:])):
```
It shouldn't matter what order delays are in and circuits are run if they delay time is stored in circuit metadata and we make the min value check more robust as suggested below

chriseclectic commented on qiskit_experiments/library/characterization/t2ramsey.py:
```diff
@@ -103,20 +103,26 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
+            raise ValueError("Delays should be strictly increasing")
 
         circuits = []
+        min_delay = self._delays[0] if self._delays[0] != 0 else self._delays[1]
```
If we make sure `self._delays` is a numpy array by doing `self._delays = np.asarray(delays)` in the init, then this can be
```suggestion
        min_delay = np.min(self._delays[self._delays > 0])
```

chriseclectic reviewed (changes requested):
Left some comments regarding improve robustness of the added code using numpy, but I don't fully understand the logic of the rotation angle change so would like one of the experimentalists to confirm its correct (@blakejohnson @wshanks @oliverdial?)

blakejohnson commented on qiskit_experiments/library/characterization/t2ramsey.py:
```diff
@@ -103,20 +103,26 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
-            circ.p(2 * np.pi * self._osc_freq, 0)
+            nosc = delay / min_delay
+            rotation_angle = 2 * np.pi * self._osc_freq * nosc
+            circ.p(rotation_angle, 0)
```
A few things:
1. I would prefer we use an `rz` gate for this. (I realize they are equivalent up to a global phase, but the `rz` really makes it explicit what you are doing.
2. In NMR they call they call this a "time proportional phase increment" (TPPI). It may be worthwhile to use that term in the doc string for the class for users more familiar with that terminology.
3. The rotation angle you want is in terms of absolute time and not relative delay, i.e.:
```python
delay_seconds = convert_unit(delay, self._unit, "s")
circ.rz(2 * np.pi * self._osc_freq * delay_seconds)
```
(assuming the existence of a `convert_unit` method that is more generic than the `convert_unit` method in terra)

With this change you can drop the book keeping about `min_delay`.

merav-aharoni replied:
@blakejohnson , thanks for the comments. I updated (1) and (3) as you suggested. It certainly makes more sense now. I will update the documentation (2) after the release.
@chriseclectic - once the tests pass (hopefully), can you review once more?

blakejohnson reviewed (changes requested):
The TPPI calculation should be referenced to absolute time.

wshanks commented:
`circuits()` looks right to me with the recent changes. Here are a few more comments I had from reading over the docstrings:

1. The units of `osc_freq` should be noted in the `__init__` docstring
2. In the `circuits()` docstring, I would rephrase "a fixed delay, a phase gate (with a linear phase)..." to "a delay, a phase gate (with phase proportional to the delay..." (any individual delay gate is fixed but the delay is being varied across circuits).
3. In the `__init__` docstring, `\lambda =2 \pi \times {osc\_freq}` should be something like `\lambda =2 \pi \times {osc\_freq} \times {delay}`.

blakejohnson commented on qiskit_experiments/library/characterization/t2ramsey.py:
```diff
@@ -116,7 +117,8 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
             circ.h(0)
             circ.delay(delay, 0, self._unit)
-            circ.p(2 * np.pi * self._osc_freq, 0)
+            rotation_angle = 2 * np.pi * self._osc_freq * delay
```
This is correct only if `delay` is in seconds. We should convert to seconds to ensure that is true.

chriseclectic replied:
I think the assumption they here is that `delay` and `osc_freq` and are same units (inverse units) so these should cancel out ie if units=s freq is Hz, but if units = us, freq is MHz

chriseclectic replied:
@merav-aharoni This assumption would break though if `units = 'dt'` so maybe best to just always make osc_freq Hz like @blakejohnson suggests

blakejohnson commented:
> The units of osc_freq should be noted in the __init__ docstring

Agreed, and it should be Hz.

nkanazawa1989 reviewed (commented):
Okey, I now see what is `osc_freq`. I prefer the word like `detuning` or `frequency_offset`, but this doesn't block merging. Actually we may need to have this value in the extra data field of frequency data entry otherwise we cannot infer true frequency offset of the signal channel from the qubit frequency.

nkanazawa1989 reviewed (approved):
LGTM
